service/history/consistency_checker.go

```diff
@@ -0,0 +1,315 @@
+		commonpb.WorkflowExecution{
+			WorkflowId: workflowKey.WorkflowID,
+			RunId:      workflowKey.RunID,
+		},
+		workflow.CallerTypeAPI,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	mutableState, err := wfContext.LoadWorkflowExecution(ctx)
+	switch err.(type) {
+	case nil:
+		if consistencyPredicate(mutableState) {
+			return newWorkflowContext(wfContext, release, mutableState), nil
+		}
+
+		wfContext.Clear()
+		mutableState, err := wfContext.LoadWorkflowExecution(ctx)
+		if err != nil {
+			release(err)
+			return nil, err
+		}
+		return newWorkflowContext(wfContext, release, mutableState), nil
+	case *serviceerror.NotFound:
+		release(err)
+		if err := assertShardOwnership(
+			ctx,
+			c.shardContext,
+			shardOwnershipAsserted,
+		); err != nil {
+			return nil, err
+		}
+		return nil, err
+	default:
+		release(err)
+		return nil, err
+	}
+}
+
+func (c *WorkflowConsistencyCheckerImpl) getCurrentWorkflowContext(
+	ctx context.Context,
+	shardOwnershipAsserted *bool,
+	consistencyPredicate MutableStateConsistencyPredicate,
+	namespaceID string,
+	workflowID string,
+) (workflowContext, error) {
+	for attempt := 1; attempt <= conditionalRetryCount; attempt++ {
+		runID, err := c.getCurrentRunID(
+			ctx,
+			shardOwnershipAsserted,
+			namespaceID,
+			workflowID,
+		)
+		if err != nil {
+			return nil, err
+		}
+		wfContext, err := c.getWorkflowContextValidatedByCheckFns(
+			ctx,
+			shardOwnershipAsserted,
+			consistencyPredicate,
+			definition.NewWorkflowKey(namespaceID, workflowID, runID),
+		)
+		if err != nil {
+			return nil, err
+		}
+		if wfContext.getMutableState().IsWorkflowExecutionRunning() {
+			return wfContext, nil
+		}
+
+		currentRunID, err := c.getCurrentRunID(
+			ctx,
+			shardOwnershipAsserted,
+			namespaceID,
+			workflowID,
+		)
+		if err != nil {
+			wfContext.getReleaseFn()(err)
+			return nil, err
+		}
+		if currentRunID == wfContext.getRunID() {
+			return wfContext, nil
+		}
+		wfContext.getReleaseFn()(nil)
+	}
+	return nil, serviceerror.NewUnavailable("unable to locate current workflow execution")
+}
+
+func (c *WorkflowConsistencyCheckerImpl) getCurrentRunID(
+	ctx context.Context,
+	shardOwnershipAsserted *bool,
+	namespaceID string,
+	workflowID string,
+) (string, error) {
+	resp, err := c.shardContext.GetCurrentExecution(
+		ctx,
+		&persistence.GetCurrentExecutionRequest{
+			ShardID:     c.shardContext.GetShardID(),
+			NamespaceID: namespaceID,
+			WorkflowID:  workflowID,
+		},
+	)
+	switch err.(type) {
+	case nil:
+		return resp.RunID, nil
+	case *serviceerror.NotFound:
+		if err := assertShardOwnership(
+			ctx,
+			c.shardContext,
+			shardOwnershipAsserted,
+		); err != nil {
+			return "", err
+		}
+		return "", err
+	default:
+		return "", err
+	}
+}
+
+func assertShardOwnership(
+	ctx context.Context,
+	shardContext shard.Context,
+	shardOwnershipAsserted *bool,
+) error {
+	if !*shardOwnershipAsserted {
+		*shardOwnershipAsserted = true
+		return shardContext.AssertOwnership(ctx)
```

Comment:
@samarabbas @paulnpdev @yiminc 
added a new shard & persistence API to assert shard ownership